topostats/io.py

```diff
@@ -388,6 +494,164 @@ def _jpk_pixel_to_nm_scaling(tiff_page: tifffile.tifffile.TiffPage) -> float:
+        Returns
+        -------
+        int
+            Size of the component in bytes.
+        """
+        component_name = read_null_terminated_string(open_file=open_file)
+        data_type = read_gwy_component_dtype(open_file=open_file)
+
+        if data_type == "o":
+            LOGGER.debug(f"component name: {component_name} | dtype: {data_type} |")
+            sub_dict = {}
+            LoadScans._gwy_read_object(open_file=open_file, data_dict=sub_dict)
+            data_dict[component_name] = sub_dict
+        elif data_type == "c":
+            value = read_char(open_file=open_file)
+            LOGGER.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
+            data_dict[component_name] = value
+        elif data_type == "i":
+            value = read_u32i(open_file=open_file)
+            LOGGER.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
+            data_dict[component_name] = value
+        elif data_type == "d":
+            value = read_64d(open_file=open_file)
+            LOGGER.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
+            data_dict[component_name] = value
+        elif data_type == "s":
+            value = read_null_terminated_string(open_file=open_file)
+            LOGGER.debug(f"component name: {component_name} | dtype: {data_type} | value: {value}")
+            data_dict[component_name] = value
+        elif data_type == "D":
+            array_size = read_u32i(open_file=open_file)
+            LOGGER.debug(f"component name: {component_name} | dtype: {data_type}")
+            LOGGER.debug(f"array size: {array_size}")
+            data = np.zeros(array_size)
+            for index in range(array_size):
+                data[index] = read_64d(open_file=open_file)
+            if "xres" in data_dict and "yres" in data_dict:
+                data = data.reshape((data_dict["xres"], data_dict["yres"]))
+            data_dict["data"] = data
+
+        return open_file.tell() - initial_byte_pos
+
+    @staticmethod
+    def _gwy_print_dict(gwy_file_dict: dict, pre_string: str) -> None:
+        """A developer function to print the nested object / component structure. Can
+        be used to find labels and values of objects / components in the `.gwy` file.
+
+        Parameters
+        ----------
+        gwy_file_dict: dict
+            Dictionary of the nested object / component structure of a `.gwy` file.
+        """
+        for key, value in gwy_file_dict.items():
+            if isinstance(value, dict):
+                print(pre_string + f"OBJECT: {key}")
+                pre_string += "  "
+                LoadScans._gwy_print_dict(gwy_file_dict=value, pre_string=pre_string)
+                pre_string = pre_string[:-2]
+            else:
+                print(pre_string + f"component: {key} | value: {value}")
+
+    @staticmethod
+    def _gwy_print_dict_wrapper(gwy_file_dict: dict) -> None:
```

Comment:
I loved this suggestion, the `json.dumps` method is so elegant, however when I went to try it...

```
000nm.0_00060.gwy
Traceback (most recent call last):
  File "/opt/homebrew/Caskroom/miniforge/base/bin/run_topostats", line 8, in <module>
    sys.exit(main())
  File "/Users/sylvi/Documents/TopoStats/topostats/run_topostats.py", line 189, in main
    all_scan_data.get_data()
  File "/Users/sylvi/Documents/TopoStats/topostats/io.py", line 705, in get_data
    self.image, self.pixel_to_nm_scaling = self.load_gwy()
  File "/Users/sylvi/Documents/TopoStats/topostats/io.py", line 649, in load_gwy
    LoadScans._gwy_print_dict(gwy_file_dict=image_data_dict, indent=2, sort_keys=False)
  File "/Users/sylvi/Documents/TopoStats/topostats/io.py", line 625, in _gwy_print_dict
    LOGGER.info(json.dumps(gwy_file_dict, indent=indent, sort_keys=sort_keys))
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 431, in _iterencode
    yield from _iterencode_dict(o, _current_indent_level)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/opt/homebrew/Caskroom/miniforge/base/lib/python3.10/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type ndarray is not JSON serializable
```

And from what I have read, each numpy array would need to be turned into a list manually, so it would require a script like I have already written, recursively turning arrays to lists...

Reply:
That is mildly annoying. I'll go through this PR again tomorrow.

Reply:
There is a neat solution to this which involves converting arrays to lists on the fly (see [SO solution](https://stackoverflow.com/a/47626762/1444043)).

Not sure that is actually any more succinct, but defining a dedicated class as in the example would make the code more generalisable and it could then be used elsewhere if required.

As above perhaps a separate issue should be created otherwise this PR will start growing legs.

Reply:
Made issue #483 👍